File: app/src/main/java/com/expansion/lg/kimaru/expansion/tables/EducationTable.java
package com.expansion.lg.kimaru.expansion.tables;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.expansion.lg.kimaru.expansion.mzigos.Education;
import com.expansion.lg.kimaru.expansion.other.Constants;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;


/**
 * Created by kimaru on 2/28/17.
 */


public class EducationTable extends SQLiteOpenHelper {

    public static final String TABLE_NAME="education";
    public static final String DATABASE_NAME= Constants.DATABASE_NAME;
    public static final int DATABASE_VERSION= Constants.DATABASE_VERSION;

    //String id, level_name, level_type, hierachy, country;

    public static String varchar_field = " varchar(512) ";
    public static String primary_field = " id INTEGER PRIMARY KEY AUTOINCREMENT ";
    public static String integer_field = " integer default 0 ";
    public static String text_field = " text ";

    public static final String ID = "_id";
    public static final String LEVEL_NAME= "name";
    public static final String TYPE = "level_type";
    public static final String HIERACHY = "hierachy";
    public static final String COUNTRY = "country";

    public static final String CREATE_DATABASE="CREATE TABLE " + TABLE_NAME + "("
            + ID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
            + LEVEL_NAME + varchar_field + ", "
            + TYPE + varchar_field + ", "
            + COUNTRY + varchar_field + ", "
            + HIERACHY + integer_field + "); ";

    String [] columns=new String[]{ID, LEVEL_NAME, TYPE, HIERACHY, COUNTRY};

    public static final String DATABASE_DROP="DROP TABLE IF EXISTS " + TABLE_NAME;

    public EducationTable(Context context) {
        super(context, TABLE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {

        db.execSQL(CREATE_DATABASE);

    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        Log.w("RegistrationTable", "upgrading database from" + oldVersion + "to" + newVersion);
        if (oldVersion < 2){
            upgradeVersion2(db);
        }
    }

    public void createEducationFromJson(JSONObject jsonObject){
        Education education = new Education();
        try {
            education.setId(jsonObject.getInt(ID));
            if (!jsonObject.isNull(LEVEL_NAME)){
                education.setLevelName(jsonObject.getString(LEVEL_NAME));
            }
            if (!jsonObject.isNull(COUNTRY)){
                education.setCountry(jsonObject.getString(COUNTRY));
            }
            if (!jsonObject.isNull(TYPE)){
                education.setLevelType(jsonObject.getString(TYPE));
            }
            if (!jsonObject.isNull(HIERACHY)){
                education.setHierachy(jsonObject.getInt(HIERACHY));
            }
            this.addEducation(education);
        }catch (Exception e){}
    }

    public long addEducation(Education education) {

        SQLiteDatabase db=getWritableDatabase();
        ContentValues cv=new ContentValues();
        cv.put(ID, education.getId());
        cv.put(LEVEL_NAME, education.getLevelName());
        cv.put(TYPE, education.getLevelType());
        cv.put(HIERACHY, education.getHierachy());
        cv.put(COUNTRY, education.getCountry());

        long id;
        if (isExist(education)){
            id = db.update(TABLE_NAME, cv, ID+"='"+education.getId()+"'", null);
        }else{
            id = db.insertWithOnConflict(TABLE_NAME, null, cv, SQLiteDatabase.CONFLICT_REPLACE);
        }
        db.close();
        return id;
    }
    public List<Education> getEducationData() {

        SQLiteDatabase db=getReadableDatabase();

        Cursor cursor=db.query(TABLE_NAME,columns,null,null,null,null,null,null);

        List<Education> educationList=new ArrayList<>();
        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            Education education=cursorToEducation(cursor);

            educationList.add(education);
        }
        db.close();
        return educationList;
    }
    public Education getEducationById(int id) {

        SQLiteDatabase db = getReadableDatabase();
        String whereClause = ID +" = ?";
        String[] whereArgs = new String[] {
                String.valueOf(id),
        };
        Cursor cursor=db.query(TABLE_NAME,columns,whereClause,whereArgs,null,null,null,null);

        if (!(cursor.moveToFirst()) || cursor.getCount() ==0){
            return null;
        }else{
            Education education=cursorToEducation(cursor);
            return education;
        }
    }

    private Education cursorToEducation(Cursor cursor){
        Education education=new Education();

        education.setId(cursor.getInt(cursor.getColumnIndex(ID)));
        education.setLevelName(cursor.getString(cursor.getColumnIndex(LEVEL_NAME)));
        education.setLevelType(cursor.getString(cursor.getColumnIndex(TYPE)));
        education.setHierachy(cursor.getInt(cursor.getColumnIndex(HIERACHY)));
        education.setCountry(cursor.getString(cursor.getColumnIndex(COUNTRY)));
        return education;
    }


    public JSONObject getEducationJson() {

        SQLiteDatabase db=getReadableDatabase();

        Cursor cursor=db.query(TABLE_NAME,columns,null,null,null,null,null,null);

        JSONObject results = new JSONObject();

        JSONArray resultSet = new JSONArray();

        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            int totalColumns = cursor.getColumnCount();
            JSONObject rowObject = new JSONObject();

            for (int i =0; i < totalColumns; i++){
                if (cursor.getColumnName(i) != null){
                    try {
                        if (cursor.getString(i) != null){
                            rowObject.put(cursor.getColumnName(i), cursor.getString(i));
                        }else{
                            rowObject.put(cursor.getColumnName(i), "");
                        }
                    }catch (Exception e){
                    }
                }
            }
            resultSet.put(rowObject);
            try {
                results.put("education", resultSet);
            } catch (JSONException e) {

            }
        }
        cursor.close();
        db.close();
        return results;
    }

    public Cursor getEducationDataCursor(String country) {

        SQLiteDatabase db=getReadableDatabase();
        String whereClause = COUNTRY+" = ?";
        String[] whereArgs = new String[] {
                country,
        };
        String [] columns=new String[]{ID, LEVEL_NAME, TYPE, HIERACHY, COUNTRY};
        Cursor cursor=db.query(TABLE_NAME,columns,whereClause,whereArgs,null,null,null,null);
        return cursor;
    }

    public void createEducationLevels(){
        //P1 to 8
        // id, hierachy;
        String levelName, levelType, country;
        // set Ugandan Education System
        for (int y=1; y <= 8; y++){
            Education education = new Education();
            education.setId(y);
            education.setCountry("UG");
            education.setHierachy(y);
            if (y == 1  ){
                education.setLevelName("Less than P7");
                education.setLevelType("primary");
            }else if(y >= 2 && y <= 7) {
                education.setLevelName("S" + (y - 1));
                education.setLevelType("secondary");
            }else{
                education.setLevelName("Tertiary");
                education.setLevelType("tertiary");
            }

            this.addEducation(education);
        }
        for (int x = 9; x <= 15; x++){
            Education education = new Education();
            education.setId(x);
            education.setHierachy(x-8);
            education.setCountry("KE");
            if (x > 14) {
                education.setLevelName("Tertiary");
                education.setLevelType("tertiary");
            }else if (x > 10){
                education.setLevelName("Form " + (x - 10));
                education.setLevelType("secondary");
            } else if (x == 10){
                education.setLevelName("P" + (x-2));
                education.setLevelType("primary");
            }else {
                education.setLevelName("Less than P" + (x-2));
                education.setLevelType("primary");
            }
            this.addEducation(education);
        }
    }

    public boolean isExist(Education education) {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cur = db.rawQuery("SELECT "+ID+" FROM " + TABLE_NAME + " WHERE "+ID+" = '" + education.getId() + "'", null);
        boolean exist = (cur.getCount() > 0);
        cur.close();
        return exist;

    }

    public void fromJson(JSONObject jsonObject){
        try{
            Education education = new Education();
            education.setId(jsonObject.getInt(ID));
            education.setLevelName(jsonObject.getString(LEVEL_NAME));
            education.setLevelType(jsonObject.getString(TYPE));
            education.setHierachy(jsonObject.getInt(HIERACHY));
            education.setCountry(jsonObject.getString(COUNTRY));
            this.addEducation(education);
        }catch (Exception e){}
    }

    private void upgradeVersion2(SQLiteDatabase db) {}


}

File: app/src/main/java/com/expansion/lg/kimaru/expansion/tables/EducationTable.kt
package com.expansion.lg.kimaru.expansion.tables

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log

import com.expansion.lg.kimaru.expansion.mzigos.Education
import com.expansion.lg.kimaru.expansion.other.Constants
import com.expansion.lg.kimaru.expansion.tables.EducationTable.Companion.TABLE_NAME

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList


/**
 * Created by kimaru on 2/28/17.
 */


class EducationTable(context: Context) : SQLiteOpenHelper(context, TABLE_NAME, null, DATABASE_VERSION) {

    internal var columns = arrayOf(ID, LEVEL_NAME, TYPE, HIERACHY, COUNTRY)
    val educationData: List<Education>
        get() {

            val db = readableDatabase

            val cursor = db.query(TABLE_NAME, columns, null, null, null, null, null, null)

            val educationList = ArrayList<Education>()
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val education = cursorToEducation(cursor)

                educationList.add(education)
                cursor.moveToNext()
            }
            db.close()
            return educationList
        }


    val educationJson: JSONObject
        get() {

            val db = readableDatabase

            val cursor = db.query(TABLE_NAME, columns, null, null, null, null, null, null)

            val results = JSONObject()

            val resultSet = JSONArray()

            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val totalColumns = cursor.columnCount
                val rowObject = JSONObject()

                for (i in 0 until totalColumns) {
                    if (cursor.getColumnName(i) != null) {
                        try {
                            if (cursor.getString(i) != null) {
                                rowObject.put(cursor.getColumnName(i), cursor.getString(i))
                            } else {
                                rowObject.put(cursor.getColumnName(i), "")
                            }
                        } catch (e: Exception) {
                        }

                    }
                }
                resultSet.put(rowObject)
                try {
                    results.put(JSON_ROOT, resultSet)
                } catch (e: JSONException) {

                }

                cursor.moveToNext()
            }
            cursor.close()
            db.close()
            return results
        }

    override fun onCreate(db: SQLiteDatabase) {

        db.execSQL(CREATE_DATABASE)

    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {

        Log.w("RegistrationTable", "upgrading database from" + oldVersion + "to" + newVersion)
        if (oldVersion < 2) {
            upgradeVersion2(db)
        }
    }

    fun createEducationFromJson(jsonObject: JSONObject) {
        val education = Education()
        try {
            education.id = jsonObject.getInt(ID)
            if (!jsonObject.isNull(LEVEL_NAME)) {
                education.levelName = jsonObject.getString(LEVEL_NAME)
            }
            if (!jsonObject.isNull(COUNTRY)) {
                education.country = jsonObject.getString(COUNTRY)
            }
            if (!jsonObject.isNull(TYPE)) {
                education.levelType = jsonObject.getString(TYPE)
            }
            if (!jsonObject.isNull(HIERACHY)) {
                education.hierachy = jsonObject.getInt(HIERACHY)
            }
            this.addEducation(education)
        } catch (e: Exception) {
        }

    }

    fun addEducation(education: Education): Long {

        val db = writableDatabase
        val cv = ContentValues()
        cv.put(ID, education.id)
        cv.put(LEVEL_NAME, education.levelName)
        cv.put(TYPE, education.levelType)
        cv.put(HIERACHY, education.hierachy)
        cv.put(COUNTRY, education.country)

        val id: Long
        if (isExist(education)) {
            id = db.update(TABLE_NAME, cv, ID + "='" + education.id + "'", null).toLong()
        } else {
            id = db.insertWithOnConflict(TABLE_NAME, null, cv, SQLiteDatabase.CONFLICT_REPLACE)
        }
        db.close()
        return id
    }

    fun getEducationById(id: Int): Education? {

        val db = readableDatabase
        val whereClause = "$ID = ?"
        val whereArgs = arrayOf(id.toString())
        val cursor = db.query(TABLE_NAME, columns, whereClause, whereArgs, null, null, null, null)

        return if (!cursor.moveToFirst() || cursor.count == 0) {
            null
        } else {
            cursorToEducation(cursor)
        }
    }

    private fun cursorToEducation(cursor: Cursor): Education {
        val education = Education()

        education.id = cursor.getInt(cursor.getColumnIndex(ID))
        education.levelName = cursor.getString(cursor.getColumnIndex(LEVEL_NAME))
        education.levelType = cursor.getString(cursor.getColumnIndex(TYPE))
        education.hierachy = cursor.getInt(cursor.getColumnIndex(HIERACHY))
        education.country = cursor.getString(cursor.getColumnIndex(COUNTRY))
        return education
    }

    fun getEducationDataCursor(country: String): Cursor {

        val db = readableDatabase
        val whereClause = "$COUNTRY = ?"
        val whereArgs = arrayOf(country)
        val columns = arrayOf(ID, LEVEL_NAME, TYPE, HIERACHY, COUNTRY)
        return db.query(TABLE_NAME, columns, whereClause, whereArgs, null, null, null, null)
    }

    fun createEducationLevels() {
        //P1 to 8
        // id, hierachy;
        val levelName: String
        val levelType: String
        val country: String
        // set Ugandan Education System
        for (y in 1..8) {
            val education = Education()
            education.id = y
            education.country = "UG"
            education.hierachy = y
            if (y == 1) {
                education.levelName = "Less than P7"
                education.levelType = "primary"
            } else if (y >= 2 && y <= 7) {
                education.levelName = "S" + (y - 1)
                education.levelType = "secondary"
            } else {
                education.levelName = "Tertiary"
                education.levelType = "tertiary"
            }

            this.addEducation(education)
        }
        for (x in 9..15) {
            val education = Education()
            education.id = x
            education.hierachy = x - 8
            education.country = "KE"
            if (x > 14) {
                education.levelName = "Tertiary"
                education.levelType = "tertiary"
            } else if (x > 10) {
                education.levelName = "Form " + (x - 10)
                education.levelType = "secondary"
            } else if (x == 10) {
                education.levelName = "P" + (x - 2)
                education.levelType = "primary"
            } else {
                education.levelName = "Less than P" + (x - 2)
                education.levelType = "primary"
            }
            this.addEducation(education)
        }
    }

    fun isExist(education: Education): Boolean {
        val db = readableDatabase
        val cur = db.rawQuery("SELECT " + ID + " FROM " + TABLE_NAME + " WHERE " + ID + " = '" + education.id + "'", null)
        val exist = cur.count > 0
        cur.close()
        return exist

    }

    fun fromJson(jsonObject: JSONObject) {
        try {
            val education = Education()
            education.id = jsonObject.getInt(ID)
            education.levelName = jsonObject.getString(LEVEL_NAME)
            education.levelType = jsonObject.getString(TYPE)
            education.hierachy = jsonObject.getInt(HIERACHY)
            education.country = jsonObject.getString(COUNTRY)
            this.addEducation(education)
        } catch (e: Exception) {
        }

    }

    private fun upgradeVersion2(db: SQLiteDatabase) {}

    companion object {

        val TABLE_NAME = "education"
        val DATABASE_NAME = Constants.DATABASE_NAME
        val DATABASE_VERSION = Constants.DATABASE_VERSION

        //String id, level_name, level_type, hierachy, country;

        var varchar_field = " varchar(512) "
        var primary_field = " id INTEGER PRIMARY KEY AUTOINCREMENT "
        var integer_field = " integer default 0 "
        var text_field = " text "

        val ID = "_id"
        val LEVEL_NAME = "name"
        val TYPE = "level_type"
        val HIERACHY = "hierachy"
        val COUNTRY = "country"
        val JSON_ROOT = "education"

        val CREATE_DATABASE = ("CREATE TABLE " + TABLE_NAME + "("
                + ID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
                + LEVEL_NAME + varchar_field + ", "
                + TYPE + varchar_field + ", "
                + COUNTRY + varchar_field + ", "
                + HIERACHY + integer_field + "); ")

        val DATABASE_DROP = "DROP TABLE IF EXISTS $TABLE_NAME"
    }


}

